apps/progress/progress.c

```diff
@@ -0,0 +1,359 @@
+#include <unistd.h>
+#include "progress.h"
+
+void insert(struct failure *f)
+{
+    struct node *node = malloc(sizeof(struct node));
+    node->val = f;
+    node->next = NULL;
+
+    if (head == NULL) {
+        head = node;
+        tail = node;
+    } else {
+        tail->next = node;
+        tail = node;
+    }
+}
+
+void free_list(struct node **head_ref)
+{
+    struct node *tmp;
+    struct node *head = *head_ref;
+    struct failure *f;
+    while (head != NULL) {
+        tmp = head;
+        head = head->next;
+        f = tmp->val;
+        free(f->name);
+        if (f->is_dynamic) free(f->error);
+        free(f);
+        free(tmp);
+    }
+}
+
+int main(int argc, char *argv[])
+{
+    for (int i = 1; i < argc; i++) {
+        char *arg = argv[i];
+        int len = strlen(arg);
+        for (int j = 0; j < len; j++) {
+            switch (arg[j]) {
+                case 'i':
+                    mode = INLINE;
+                    break;
+                case 'c':
+                    mode = CHARACTER;
+                    break;
+                case 'l':
+                    mode = LINES;
+                    break;
+                case 's':
+                    mode = SUMMARY;
+                    break;
+                case 't':
+                    indent++;
+                    break;
+            }
+        }
+    }
+
+    if (mode == SUMMARY) indent = 0;
+
+    report_start();
+    while (1) {
+        char v[1];
+        if (read(STDIN, v, 1) == 0) {
+            process_line();
+            break;
+        }
+
+        char b = v[0];
+        if (b == CTRL_C) break;
+        if (b == CTRL_D) break;
+        if (b == '\n') {
+            process_line();
+        } else {
+            if (line_end < 4096) line_buffer[line_end++] = b;
+        }
+    }
+    return report_finish();
+}
+
+void process_line()
+{
+    int pos = 3, end = line_end;
+    if (line_end < pos) return;
+    if (line_buffer[0] == '#' && line_buffer[1] == '#') {
+        switch (line_buffer[2]) {
+            case '+': ;
+                // begin the next test
+                char *test_name = str_slice(line_buffer, pos, end);
+                report_test_begin(test_name);
+                break;
+            case '-': ;
+                // end the current test
+                int passed = match_str("ok", line_buffer, pos, end);
+                if (passed) report_test_passed();
+                else report_test_failed(str_slice(line_buffer, pos, end), 1);
+                break;
+            case '>': ;
+                // increase the total test count
+                char *count_str = str_slice(line_buffer, pos, end);
+                int count = strtol(count_str, NULL, 10);
+                if (count > 0) test_count += count;
+                free(count_str);
+                break;
+        }
+    }
+    line_end = 0;
+}
+
+void outc(char c)
+{
+    printf("%c", c);
+    char_backup++;
+}
+
+void outi(int i)
+{
+    printf("%d", i);
```

Comment:
printf returns the number of characters printed, so you don't have to calculate it manually.